File: itk/agents/go/v10/main.go
package main

import (
	"context"
	"encoding/base64"
	"encoding/json"
	"flag"
	"fmt"
	"iter"
	"log"
	"net"
	"net/http"
	"os"
	"os/signal"
	"strings"
	"syscall"
	"time"

	"itk/agents/go/v10/pb"

	"github.com/a2aproject/a2a-go/a2a"
	"github.com/a2aproject/a2a-go/a2aclient"
	"github.com/a2aproject/a2a-go/a2aclient/agentcard"
	"github.com/a2aproject/a2a-go/a2acompat/a2av0"
	a2agrpc "github.com/a2aproject/a2a-go/a2agrpc/v1"
	"github.com/a2aproject/a2a-go/a2asrv"
	"golang.org/x/sync/errgroup"
	"google.golang.org/grpc"
	"google.golang.org/grpc/credentials/insecure"
	"google.golang.org/protobuf/proto"
)

type V10AgentExecutor struct{}

func (e *V10AgentExecutor) Execute(ctx context.Context, execCtx *a2asrv.ExecutorContext) iter.Seq2[a2a.Event, error] {
	return func(yield func(a2a.Event, error) bool) {
		log.Printf("Executing task %s", string(execCtx.TaskID))

		if execCtx.StoredTask == nil {
			if !yield(a2a.NewSubmittedTask(execCtx, execCtx.Message), nil) {
				return
			}
		}

		if !yield(a2a.NewStatusUpdateEvent(execCtx, a2a.TaskStateWorking, nil), nil) {
			return
		}

		instruction, err := extractInstruction(execCtx.Message)
		if err != nil {
			log.Printf("Error: %v", err)
			yield(a2a.NewMessage(a2a.MessageRoleAgent, a2a.NewTextPart(err.Error())), nil)
			return
		}

		results, err := e.handleInstruction(ctx, execCtx, instruction)
		if err != nil {
			log.Printf("Error handling instruction: %v", err)
			yield(a2a.NewStatusUpdateEvent(execCtx, a2a.TaskStateFailed, nil), nil)
			return
		}

		response := strings.Join(results, "\n")
		if !yield(a2a.NewStatusUpdateEvent(execCtx, a2a.TaskStateCompleted, a2a.NewMessage(a2a.MessageRoleAgent, a2a.NewTextPart(response))), nil) {
			return
		}
	}
}

func extractInstruction(msg *a2a.Message) (*pb.Instruction, error) {
	for _, part := range msg.Parts {
		if part.MediaType == "application/x-protobuf" || (part.MediaType == "" && part.Filename == "instruction.bin") {
			raw := part.Raw()
			if len(raw) > 0 {
				var instruction pb.Instruction
				if err := proto.Unmarshal(raw, &instruction); err == nil {
					return &instruction, nil
				}
			}
		}
		text := part.Text()
		if text != "" {
			if raw, err := base64.StdEncoding.DecodeString(text); err == nil {
				var instruction pb.Instruction
				if err := proto.Unmarshal(raw, &instruction); err == nil {
					return &instruction, nil
				}
			}
		}
	}
	return nil, fmt.Errorf("no valid instruction found in request")
}

func (e *V10AgentExecutor) handleInstruction(ctx context.Context, execCtx *a2asrv.ExecutorContext, inst *pb.Instruction) ([]string, error) {
	switch {
	case inst.GetCallAgent() != nil:
		return e.handleCallAgent(ctx, inst.GetCallAgent())

	case inst.GetReturnResponse() != nil:
		return []string{inst.GetReturnResponse().Response}, nil

	case inst.GetSteps() != nil:
		var allResults []string
		for _, step := range inst.GetSteps().Instructions {
			results, err := e.handleInstruction(ctx, execCtx, step)
			if err != nil {
				return nil, err
			}
			allResults = append(allResults, results...)
		}
		return allResults, nil

	default:
		return nil, fmt.Errorf("unknown instruction type")
	}
}

func (e *V10AgentExecutor) handleCallAgent(ctx context.Context, call *pb.CallAgent) ([]string, error) {
	log.Printf("Calling agent %s via %s", call.AgentCardUri, call.Transport)

	// 1. Resolve agent card
	resolver := agentcard.NewResolver(nil)
	resolver.CardParser = a2av0.NewAgentCardParser()
	card, err := resolver.Resolve(ctx, call.AgentCardUri)
	if err != nil {
		return nil, fmt.Errorf("failed to resolve agent card for %s: %w", call.AgentCardUri, err)
	}

	// Print parsed card for debugging as requested
	if cardJSON, mErr := json.MarshalIndent(card, "", "  "); mErr == nil {
		log.Printf("Parsed Agent Card for %s:\n%s", call.AgentCardUri, string(cardJSON))
	} else {
		log.Printf("Warning: failed to marshal agent card for logging: %v", mErr)
	}

	protocol := mapTransport(call.Transport)
	log.Printf("Mapped transport: %s", protocol)

	// 3. Find all matching interfaces from the card
	matchedInterfaces := selectInterfaces(protocol, card)
	if len(matchedInterfaces) == 0 {
		return nil, fmt.Errorf("transport protocol %s is not supported by agent %s", protocol, call.AgentCardUri)
	}

	// 4. Create client using a factory
	// We instantiate a client through factory.CreateFromEndpoints
	// to strictly enforce the transport protocol. a2aclient.NewFromCard
	// seems to be using the first available transport if the specified one
	// is not supported.
	var factory *a2aclient.Factory
	compatFactory := a2av0.NewJSONRPCTransportFactory(a2av0.JSONRPCTransportConfig{})
	factory = a2aclient.NewFactory(
		a2agrpc.WithGRPCTransport(grpc.WithTransportCredentials(insecure.NewCredentials())),
		a2aclient.WithCompatTransport("0.3", a2a.TransportProtocolJSONRPC, compatFactory),
		a2aclient.WithCompatTransport("", a2a.TransportProtocolJSONRPC, compatFactory),
	)

	client, err := factory.CreateFromEndpoints(ctx, matchedInterfaces)
	if err != nil {
		return nil, fmt.Errorf("failed to create client: %w", err)
	}

	wrappedMsg, err := wrapInstructionToRequest(call.Instruction)
	if err != nil {
		return nil, fmt.Errorf("failed to wrap nested instruction: %w", err)
	}

	var responses []string
	if call.Streaming {
		events := client.SendStreamingMessage(ctx, &a2a.SendMessageRequest{Message: wrappedMsg})
		for ev, err := range events {
			if err != nil {
				log.Printf("Error inside streaming call to %s: %v", call.AgentCardUri, err)
				return nil, fmt.Errorf("streaming call failed to agent %s: %w", call.AgentCardUri, err)
			}
			responses = append(responses, extractResponses(ev)...)
		}
	} else {
		result, err := client.SendMessage(ctx, &a2a.SendMessageRequest{
			Message: wrappedMsg,
		})
		if err != nil {
			log.Printf("Error sending message to %s: %v", call.AgentCardUri, err)
			return nil, fmt.Errorf("failed to send message to agent %s: %w", call.AgentCardUri, err)
		}
		responses = extractResponses(result)
	}

	log.Printf("Received responses from %s", call.AgentCardUri)
	return responses, nil
}

func extractResponses(result any) []string {
	var responses []string
	log.Printf("Extracting responses from result of type %T", result)
	switch r := result.(type) {
	case *a2a.Message:
		for _, part := range r.Parts {
			if t := part.Text(); t != "" {
				responses = append(responses, t)
			}
		}
	case *a2a.Task:
		// Check both Status.Message and History
		if r.Status.Message != nil {
			for _, part := range r.Status.Message.Parts {
				if t := part.Text(); t != "" {
					responses = append(responses, t)
				}
			}
		}
		for _, msg := range r.History {
			if msg.Role == a2a.MessageRoleAgent {
				for _, part := range msg.Parts {
					if t := part.Text(); t != "" {
						responses = append(responses, t)
					}
				}
			}
		}
	case *a2a.TaskStatusUpdateEvent:
		if r.Status.Message != nil {
			for _, part := range r.Status.Message.Parts {
				if t := part.Text(); t != "" {
					responses = append(responses, t)
				}
			}
		}
	default:
		log.Printf("Warning: unexpected result type from SendMessage: %T", result)
	}
	return responses
}

func (e *V10AgentExecutor) Cancel(_ context.Context, execCtx *a2asrv.ExecutorContext) iter.Seq2[a2a.Event, error] {
	return func(yield func(a2a.Event, error) bool) {
		log.Printf("Cancel requested for task %s", string(execCtx.TaskID))
		yield(a2a.NewStatusUpdateEvent(execCtx, a2a.TaskStateCanceled, nil), nil)
	}
}

func wrapInstructionToRequest(inst *pb.Instruction) (*a2a.Message, error) {
	instBytes, err := proto.Marshal(inst)
	if err != nil {
		return nil, err
	}

	part := a2a.NewRawPart(instBytes)
	part.Filename = "instruction.bin"
	part.MediaType = "application/x-protobuf"

	return a2a.NewMessage(a2a.MessageRoleUser, part), nil
}

func mapTransport(t string) a2a.TransportProtocol {
	switch strings.ToUpper(t) {
	case "GRPC":
		return a2a.TransportProtocolGRPC
	case "REST", "HTTP_JSON", "HTTP+JSON":
		return a2a.TransportProtocolHTTPJSON
	default:
		return a2a.TransportProtocolJSONRPC
	}
}


func selectInterfaces(protocol a2a.TransportProtocol, card *a2a.AgentCard) []*a2a.AgentInterface {
	var matched []*a2a.AgentInterface
	for _, iface := range card.SupportedInterfaces {
		if iface.ProtocolBinding == protocol {
			iface.URL = strings.TrimSuffix(iface.URL, "/")
			matched = append(matched, iface)
		}
	}
	return matched
}

var httpPort = flag.Int("httpPort", 10102, "HTTP port")
var grpcPort = flag.Int("grpcPort", 11002, "gRPC port")


func main() {
	if err := run(); err != nil {
		log.Fatalf("Server session ended with error: %v", err)
	}
}

func run() error {
	flag.Parse()

	jsonRPCV0Addr := fmt.Sprintf("http://127.0.0.1:%d", *httpPort)

	agentCard := &a2a.AgentCard{
		Name:               "ITK v10 Agent",
		Description:        "Multi-transport Go agent with A2A v0.3 compatibility.",
		Version:            "1.0.0-alpha",
		Capabilities:       a2a.AgentCapabilities{Streaming: true},
		DefaultInputModes:  []string{"text"},
		DefaultOutputModes: []string{"text"},
		SupportedInterfaces: []*a2a.AgentInterface{
			{
				URL:             fmt.Sprintf("http://127.0.0.1:%d/jsonrpc", *httpPort),
				ProtocolBinding: a2a.TransportProtocolJSONRPC,
				ProtocolVersion: a2a.Version,
			},
			{
				URL:             jsonRPCV0Addr,
				ProtocolBinding: a2a.TransportProtocolJSONRPC,
				ProtocolVersion: a2av0.Version,
			},
			{
				URL:             fmt.Sprintf("http://127.0.0.1:%d/rest", *httpPort),
				ProtocolBinding: a2a.TransportProtocolHTTPJSON,
				ProtocolVersion: a2a.Version,
			},
			{
				URL:             fmt.Sprintf("127.0.0.1:%d", *grpcPort),
				ProtocolBinding: a2a.TransportProtocolGRPC,
				ProtocolVersion: a2a.Version,
			},
		},
	}

	executor := &V10AgentExecutor{}
	requestHandler := a2asrv.NewHandler(executor)

	// Servers
	mux := http.NewServeMux()
	agentCardRoute := fmt.Sprintf("/jsonrpc%s", a2asrv.WellKnownAgentCardPath)
	mux.Handle("/", a2av0.NewJSONRPCHandler(requestHandler))
	mux.Handle("/jsonrpc", a2asrv.NewJSONRPCHandler(requestHandler))
	mux.Handle("/rest/", http.StripPrefix("/rest", a2asrv.NewRESTHandler(requestHandler)))

	cardProducer := a2av0.NewStaticAgentCardProducer(agentCard)
	mux.Handle(agentCardRoute, a2asrv.NewAgentCardHandler(cardProducer))

	httpServer := &http.Server{
		Addr:              fmt.Sprintf(":%d", *httpPort),
		Handler:           mux,
		ReadHeaderTimeout: 3 * time.Second,
	}

	ctx, stop := signal.NotifyContext(context.Background(), os.Interrupt, syscall.SIGTERM)
	defer stop()

	g, ctx := errgroup.WithContext(ctx)

	g.Go(func() error {
		serverType := "consolidated v1.0 & v0.3"
		log.Printf("Starting HTTP server on 127.0.0.1:%d (%s)", *httpPort, serverType)
		if err := httpServer.ListenAndServe(); err != nil && err != http.ErrServerClosed {
			return err
		}
		return nil
	})

	grpcServer := grpc.NewServer()
	a2agrpc.NewHandler(requestHandler).RegisterWith(grpcServer)
	g.Go(func() error {
		lis, err := net.Listen("tcp", fmt.Sprintf(":%d", *grpcPort))
		if err != nil {
			return err
		}
		log.Printf("Starting gRPC server on 127.0.0.1:%d", *grpcPort)
		return grpcServer.Serve(lis)
	})
	g.Go(func() error {
		<-ctx.Done()
		grpcServer.GracefulStop()
		return nil
	})

	// Wait for stop signal
	g.Go(func() error {
		<-ctx.Done()
		log.Println("Shutting down servers...")
		shutdownCtx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
		defer cancel()
		return httpServer.Shutdown(shutdownCtx)
	})

	return g.Wait()
}

File: itk/agents/python/v10/main.py
import argparse  # noqa: I001
import asyncio
import base64
import logging
import uuid

import grpc
import httpx
import uvicorn

from fastapi import FastAPI

from pyproto import instruction_pb2

from a2a.client import ClientConfig, ClientFactory
from a2a.compat.v0_3 import a2a_v0_3_pb2_grpc
from a2a.compat.v0_3.grpc_handler import CompatGrpcHandler
from a2a.server.agent_execution import AgentExecutor, RequestContext
from a2a.server.apps import A2AFastAPIApplication, A2ARESTFastAPIApplication
from a2a.server.events import EventQueue
from a2a.server.events.in_memory_queue_manager import InMemoryQueueManager
from a2a.server.request_handlers import DefaultRequestHandler, GrpcHandler
from a2a.server.tasks import TaskUpdater
from a2a.server.tasks.inmemory_task_store import InMemoryTaskStore
from a2a.types import a2a_pb2_grpc
from a2a.types.a2a_pb2 import (
    AgentCapabilities,
    AgentCard,
    AgentInterface,
    Message,
    Part,
    SendMessageRequest,
    TaskState,
)
from a2a.utils import TransportProtocol


logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)


def extract_instruction(
    message: Message | None,
) -> instruction_pb2.Instruction | None:
    """Extracts an Instruction proto from an A2A Message."""
    if not message or not message.parts:
        return None

    for part in message.parts:
        # 1. Handle binary protobuf part (media_type or filename)
        if (
            part.media_type == 'application/x-protobuf'
            or part.filename == 'instruction.bin'
        ):
            try:
                inst = instruction_pb2.Instruction()
                if part.raw:
                    inst.ParseFromString(part.raw)
                elif part.text:
                    # Some clients might send it as base64 in text part
                    raw = base64.b64decode(part.text)
                    inst.ParseFromString(raw)
            except Exception:  # noqa: BLE001
                logger.debug(
                    'Failed to parse instruction from binary part',
                    exc_info=True,
                )
                continue
            else:
                return inst

        # 2. Handle base64 encoded instruction in any text part
        if part.text:
            try:
                raw = base64.b64decode(part.text)
                inst = instruction_pb2.Instruction()
                inst.ParseFromString(raw)
            except Exception:  # noqa: BLE001
                logger.debug(
                    'Failed to parse instruction from text part', exc_info=True
                )
                continue
            else:
                return inst
    return None


def wrap_instruction_to_request(inst: instruction_pb2.Instruction) -> Message:
    """Wraps an Instruction proto into an A2A Message."""
    inst_bytes = inst.SerializeToString()
    return Message(
        role='ROLE_USER',
        message_id=str(uuid.uuid4()),
        parts=[
            Part(
                raw=inst_bytes,
                media_type='application/x-protobuf',
                filename='instruction.bin',
            )
        ],
    )


async def handle_call_agent(call: instruction_pb2.CallAgent) -> list[str]:
    """Handles the CallAgent instruction by invoking another agent."""
    logger.info('Calling agent %s via %s', call.agent_card_uri, call.transport)

    # Mapping transport string to TransportProtocol enum
    transport_map = {
        'JSONRPC': TransportProtocol.JSONRPC,
        'HTTP+JSON': TransportProtocol.HTTP_JSON,
        'HTTP_JSON': TransportProtocol.HTTP_JSON,
        'REST': TransportProtocol.HTTP_JSON,
        'GRPC': TransportProtocol.GRPC,
    }

    selected_transport = transport_map.get(
        call.transport.upper(), TransportProtocol.JSONRPC
    )
    if selected_transport is None:
        raise ValueError(f'Unsupported transport: {call.transport}')

    config = ClientConfig()
    config.httpx_client = httpx.AsyncClient(timeout=30.0)
    config.grpc_channel_factory = grpc.aio.insecure_channel
    config.supported_protocol_bindings = [selected_transport]
    config.streaming = call.streaming or (
        selected_transport == TransportProtocol.GRPC
    )

    try:
        client = await ClientFactory.connect(
            call.agent_card_uri,
            client_config=config,
        )

        # Wrap nested instruction
        nested_msg = wrap_instruction_to_request(call.instruction)
        request = SendMessageRequest(message=nested_msg)

        results = []
        async for event in client.send_message(request):
            # Event is streaming response and task
            logger.info('Event: %s', event)
            stream_resp, task = event

            message = None
            if stream_resp.HasField('message'):
                message = stream_resp.message
            elif task and task.status.HasField('message'):
                message = task.status.message
            elif stream_resp.HasField(
                'status_update'
            ) and stream_resp.status_update.status.HasField('message'):
                message = stream_resp.status_update.status.message

            if message:
                results.extend(part.text for part in message.parts if part.text)

    except Exception as e:
        logger.exception('Failed to call outbound agent')
        raise RuntimeError(
            f'Outbound call to {call.agent_card_uri} failed: {e!s}'
        ) from e
    else:
        return results


async def handle_instruction(inst: instruction_pb2.Instruction) -> list[str]:
    """Recursively handles instructions."""
    if inst.HasField('call_agent'):
        return await handle_call_agent(inst.call_agent)
    if inst.HasField('return_response'):
        return [inst.return_response.response]
    if inst.HasField('steps'):
        all_results = []
        for step in inst.steps.instructions:
            results = await handle_instruction(step)
            all_results.extend(results)
        return all_results
    raise ValueError('Unknown instruction type')


class V10AgentExecutor(AgentExecutor):
    """Executor for ITK v10 agent tasks."""

    async def execute(
        self, context: RequestContext, event_queue: EventQueue
    ) -> None:
        """Executes a task instruction."""
        logger.info('Executing task %s', context.task_id)
        task_updater = TaskUpdater(
            event_queue,
            context.task_id,
            context.context_id,
        )

        await task_updater.update_status(TaskState.TASK_STATE_SUBMITTED)
        await task_updater.update_status(TaskState.TASK_STATE_WORKING)

        instruction = extract_instruction(context.message)
        if not instruction:
            error_msg = 'No valid instruction found in request'
            logger.error(error_msg)
            await task_updater.update_status(
                TaskState.TASK_STATE_FAILED,
                message=task_updater.new_agent_message([Part(text=error_msg)]),
            )
            return

        try:
            logger.info('Instruction: %s', instruction)
            results = await handle_instruction(instruction)
            response_text = '\n'.join(results)
            logger.info('Response: %s', response_text)
            await task_updater.update_status(
                TaskState.TASK_STATE_COMPLETED,
                message=task_updater.new_agent_message(
                    [Part(text=response_text)]
                ),
            )
            logger.info('Task %s completed', context.task_id)
        except Exception as e:
            logger.exception('Error during instruction handling')
            await task_updater.update_status(
                TaskState.TASK_STATE_FAILED,
                message=task_updater.new_agent_message([Part(text=str(e))]),
            )

    async def cancel(
        self, context: RequestContext, event_queue: EventQueue
    ) -> None:
        """Cancels a task."""
        logger.info('Cancel requested for task %s', context.task_id)
        task_updater = TaskUpdater(
            event_queue,
            context.task_id,
            context.context_id,
        )
        await task_updater.update_status(TaskState.TASK_STATE_CANCELED)


async def main_async(http_port: int, grpc_port: int) -> None:
    """Starts the Agent with HTTP and gRPC interfaces."""
    interfaces = [
        AgentInterface(
            protocol_binding=TransportProtocol.GRPC,
            url=f'127.0.0.1:{grpc_port}',
            protocol_version='1.0',
        ),
        AgentInterface(
            protocol_binding=TransportProtocol.GRPC,
            url=f'127.0.0.1:{grpc_port}',
            protocol_version='0.3',
        ),
    ]

    interfaces.append(
        AgentInterface(
            protocol_binding=TransportProtocol.JSONRPC,
            url=f'http://127.0.0.1:{http_port}/jsonrpc/',
        )
    )
    interfaces.append(
        AgentInterface(
            protocol_binding=TransportProtocol.HTTP_JSON,
            url=f'http://127.0.0.1:{http_port}/rest/',
            protocol_version='1.0',
        )
    )
    interfaces.append(
        AgentInterface(
            protocol_binding=TransportProtocol.HTTP_JSON,
            url=f'http://127.0.0.1:{http_port}/rest/',
            protocol_version='0.3',
        )
    )

    agent_card = AgentCard(
        name='ITK v10 Agent',
        description='Python agent using SDK 1.0.',
        version='1.0.0',
        capabilities=AgentCapabilities(streaming=True),
        default_input_modes=['text/plain'],
        default_output_modes=['text/plain'],
        supported_interfaces=interfaces,
    )

    task_store = InMemoryTaskStore()
    handler = DefaultRequestHandler(
        agent_executor=V10AgentExecutor(),
        task_store=task_store,
        queue_manager=InMemoryQueueManager(),
    )

    app = FastAPI()

    json_rpc_app = A2AFastAPIApplication(
        agent_card, handler, enable_v0_3_compat=True
    ).build()
    app.mount('/jsonrpc', json_rpc_app)
    rest_app = A2ARESTFastAPIApplication(
        http_handler=handler, agent_card=agent_card, enable_v0_3_compat=True
    ).build()
    app.mount('/rest', rest_app)

    server = grpc.aio.server()

    compat_servicer = CompatGrpcHandler(agent_card, handler)
    a2a_v0_3_pb2_grpc.add_A2AServiceServicer_to_server(compat_servicer, server)
    servicer = GrpcHandler(agent_card, handler)
    a2a_pb2_grpc.add_A2AServiceServicer_to_server(servicer, server)

    server.add_insecure_port(f'127.0.0.1:{grpc_port}')
    await server.start()

    logger.info(
        'Starting ITK v10 Agent on HTTP port %s and gRPC port %s',
        http_port,
        grpc_port,
    )

    config = uvicorn.Config(
        app, host='127.0.0.1', port=http_port, log_level='info'
    )
    uvicorn_server = uvicorn.Server(config)

    await uvicorn_server.serve()


def str2bool(v: str | bool) -> bool:
    """Converts a string to a boolean value."""
    if isinstance(v, bool):
        return v
    if v.lower() in ('yes', 'true', 't', 'y', '1'):
        return True
    if v.lower() in ('no', 'false', 'f', 'n', '0'):
        return False
    raise argparse.ArgumentTypeError('Boolean value expected.')


def main() -> None:
    """Main entry point for the agent."""
    parser = argparse.ArgumentParser()
    parser.add_argument('--httpPort', type=int, default=10102)
    parser.add_argument('--grpcPort', type=int, default=11002)
    args = parser.parse_args()

    asyncio.run(main_async(args.httpPort, args.grpcPort))


if __name__ == '__main__':
    main()
